operator/api/redpanda/v1alpha3/redpanda_conversion.go

```diff
@@ -0,0 +1,954 @@
+package v1alpha3
+
+import (
+	"encoding/json"
+	"fmt"
+	"maps"
+	"reflect"
+	"slices"
+	"strings"
+	"time"
+	"unsafe"
+
+	"github.com/cockroachdb/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	applycorev1 "k8s.io/client-go/applyconfigurations/core/v1"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+	"sigs.k8s.io/yaml"
+
+	"github.com/redpanda-data/redpanda-operator/charts/redpanda/v5"
+	redpandav1alpha2 "github.com/redpanda-data/redpanda-operator/operator/api/redpanda/v1alpha2"
+	vectorizedv1alpha1 "github.com/redpanda-data/redpanda-operator/operator/api/vectorized/v1alpha1"
+)
+
+var _ conversion.Convertible = (*Redpanda)(nil)
+
+func (dst *Redpanda) ConvertFrom(raw conversion.Hub) (err error) { //nolint:staticcheck // changing receiver names makes this much more clear.
+	// To make writing conversion code less tedious, convertFrom panics instead
+	// of handling errors. We catch them here and convert them to errors.
+	defer func() {
+		switch r := recover().(type) {
+		case nil:
+		case error:
+			err = errors.Wrap(r, "ConvertFrom failed")
+		default:
+			err = errors.Newf("ConvertFrom failed with panic: %v", r)
+		}
+	}()
+
+	src, ok := raw.(*redpandav1alpha2.Redpanda)
+	if !ok {
+		return errors.Newf("expected dst to be %T; got %T", (*redpandav1alpha2.Redpanda)(nil), raw)
+	}
+
+	values, err := src.GetValues()
+	if err != nil {
+		return err
+	}
+
+	*dst = convertFrom[Redpanda](src, &values).(Redpanda)
+	return nil
+}
+
+func (src *Redpanda) ConvertTo(raw conversion.Hub) (err error) { //nolint:staticcheck // changing receiver names makes this much more clear.
+	// To make writing conversion code less tedious, convertTo panics instead
+	// of handling errors. We catch them here and convert them to errors.
+	defer func() {
+		switch r := recover().(type) {
+		case nil:
+		case error:
+			err = errors.Wrap(r, "ConvertTo failed")
+		default:
+			err = errors.Newf("ConvertTo failed with panic: %v", r)
+		}
+	}()
+
+	dst, ok := raw.(*redpandav1alpha2.Redpanda)
+	if !ok {
+		return errors.Newf("expected dst to be %T; got %T", (*redpandav1alpha2.Redpanda)(nil), raw)
+	}
+
+	*dst = convertTo[redpandav1alpha2.Redpanda](src).(redpandav1alpha2.Redpanda)
+	return nil
+}
+
+func convertFrom[T any](src *redpandav1alpha2.Redpanda, values *redpanda.Values) any {
+	t := reflect.New(reflect.TypeFor[T]()).Elem().Interface()
+
+	switch t.(type) {
+	case Redpanda:
+		return Redpanda{
+			TypeMeta:   metav1.TypeMeta{APIVersion: "", Kind: ""},
+			ObjectMeta: *src.ObjectMeta.DeepCopy(),
+			Spec:       convertFrom[RedpandaSpec](src, values).(RedpandaSpec),
+			Status:     convertFrom[RedpandaStatus](src, values).(RedpandaStatus),
+		}
+
+	case RedpandaSpec:
+		return RedpandaSpec{
+			ClusterDomain: ptr.Deref(src.Spec.ClusterSpec.ClusterDomain, ""),
```

Comment:
When we're delegating to `src` v. `values` seems inconsistent. Is it b/c we don't want to set based on defaults in some places, but do others? Mind calling out where we do/don't some places?

Reply:
Added a top level comment:
```
	// Throughout this function src and values are both referenced using the
	// general mantra: If a value is _truly_ optional, source it from src to
	// preserve it's presence or lack thereof. If a value is required, source
	// it from values to fallback to the default.
	// e.g. Replicas is technically optional as they're nilable on STS's. Where
	// as the RPC listener port is not.
	// The rapid tests, do a surprisingly good good of requiring this behavior
	// to be upheld.
```

Most instances you see of it going one way or the other are actually informed by the rapid tests more so than me making an informed decision 😅